clay: add safe merge strategies

philipcmonk opened the pull request:
See gen/hood/merge/help.txt for an overview of all merge strategies.

- `|merge` with no arguments now prints out usage and a description of merge strategies
- Added `%take-this` and `%take-that` strategies, which are the same as the old `%this` and `%that` except they preserve files which only exist on one side or the other.  The primary use-case for `%take-that` is "forcing" updates without obliterating locally-added files, such as notebooks or 3rd party apps.
- Added `%meet-this` and `%meet-that` strategies, which are the same as `%meet`, except if both sides touched the same file (since their most recent common ancestor), prefer this or that desk.  Note this operates on the file level -- even if the changes inside the file don't conflict, we take the whole file from one side or the other.  It would be nice to add `%mate-this` and `%mate-that` strategies which operate within the file.
- Renamed `%this` and `%that` to `%only-this` and `%only-that` to emphasize that they remove anything not in the other desk.
- Note `%take-this` and `%take-that` are 2-way merges (like `%only-this` and `%only-that`).  This means they only look at the content in the commits, not their ancestry.  This means they are suitable for cases where you don't have a unique mergebase (`%merge-no-merge-base` and `%criss-cross`) and generally makes them blunt weapons.  By contrast, `%meet-this` and `%meet-that` *do* require a mergebase and are a sort of 3-way merge.
- Switch the default merge strategy for desks on their first revision from `%that` to `%take-that`.  This should alleviate some issues where people add files (eg make a notebook) before the first OTA arrives.

belisarius222 commented on pkg/arvo/gen/hood/merge/help.txt:
```diff
@@ -0,0 +1,110 @@
+of a file.  Identical to %meet, except that some marks, like %hoon,
+allow intelligent merge of changes to different parts of a file.
+                                                                      
+%meld: combine changes; if both sides changed the same part of a
```
Does this mean %meld doesn't apply either change, and just goes with the old version?

philipcmonk replied:
That's correct.  This makes it convenient to construct a 3-way diff later on, but it's somewhat non-intuitive

belisarius222 replied:
Ah, I see, this is used for annotating conflicts on top of it.  Makes sense.

belisarius222 commented on pkg/arvo/gen/hood/merge/help.txt:
```diff
@@ -0,0 +1,110 @@
+any files you separately added.
+                                                                      
+We speak of merging into a destination *desk* from a source *commit*
+becaus while you can only merge on top of a desk, you can merge from
```
typo: becaus

belisarius222 commented on pkg/arvo/gen/hood/merge/help.txt:
```diff
@@ -0,0 +1,110 @@
+                                                                      
+If you resolve merge conflicts manually, for example by mounting the
+desks, copying the files in unix and then running |commit, you
+should usually run an %only-this merge.  This will not change the
```
As a note for the future, I think we should eventually automate this workflow for development.  It could be related to the `-work` thread.

philipcmonk replied:
Good idea.  Would love it to to just pop up your `$EDITOR`, but that's probably a bridge too far...

belisarius222 reviewed (approved):
LGTM other than that one typo

philipcmonk commented:
Fixed, and also fixed a small bug in `%mate` that I noticed while trying to update a 3rd party app: https://github.com/urbit/urbit/pull/3404/commits/c7480c9b0700274afdb92a009fc01f53ba08e9fb
